src/System.Device.Gpio/System/Device/Pwm/Channels/UnixPwmChannel.Linux.cs

```diff
@@ -96,12 +113,16 @@ private void SetFrequency(int frequency)
             {
                 throw new ArgumentOutOfRangeException(nameof(frequency), frequency, "Value must not be negative.");
             }
-
+            double oldDutyCycle = _dutyCycle;
+            // We have to set the duty cycle to 0 before changing the frequency, otherwise an InvalidArgumentException occurs when writing the period file and the 
+            // current duty cycle is longer than the period. We have to recompute the duty cycle (in nanoseconds) after this, anyway
+            SetDutyCycle(0);
```

Comment:
This is only the case when frequency will be less than dutycycle, so we don't want to set DutyCycle to 0 every time. What needs to be fixed here that we are not doing is to recalculate the DutyCycle after changing the frequency. This is a bug since today what we do is to set a DutyCycle which is a percentage of the frequency, but when we update the frequency we don't seem to update the DutyCycle so that is a bug.

Reply:
That part is fixed with my change as well (as I'm re-setting the existing percentage afterwards). Since the duty cycle is incorrect when changing the freqency in either case and I'm assuming that one wouldn't normally change the frequency all the time, I don't think we should complicate the logic much further and distinguish whether we increase or decrease the frequency or the duty cycle.